### ISSUE TITLE
Purpose of this package: API or I-feel-lucky Plone Solr integration?

### ISSUE THREAD

**Issue by tomgross** [collective/collective.solr#135]
While polishing this package for a Plone 5 compatible release I found myself asking the purpose of this package. Should it be

1. an API for plugging in other (UI) products or
2. a full fledged solution for all relevant Plone-Solr integration tasks?

I'm tending to go for 1. because of the following reasons:

With an interface to Solr and a complex (react based) UI we have many dependencies which are hard 

- to track. We need to make sure newer versions of Solr and UI components are supported and tested.
- to test. We have low level network tests side-by-side with UI tests.
- for new people to start with. You have to understand Plone, Solr and React based UIs
- Searches are mostly custom tailerd. It will we hard to maintain a one-size-fits-all-approach.

I suggest to focus on a stable API and do the following for the next major release:

  1. Remove the complex React based search UI and put it in a seperate product. I myself don't use it because I work with eea.facetednavigation and there might be other scenarios.
  2. (Re)add a simple BrowserView based search form using the product with some (faceting) enhancements. This is far more easier to test across diferent Plone versions!
 3. Document how to override Plones search UI with a custom one in Plone 4.3 and Plone 5.

Any feedback? @tisto @gforcada

**Comment by tisto** [comment]
I asked myself the same question before I went ahead with the React
solution. I actually started to implement different jquery/react/angular
solution as add-on products.

Though, bottom line is that c.solr should provide a good out-of-the-box
solution and if we spread functionality across packages we will never
have one good solution. People will have to try different solutions
which will most likely be a frustrating experience. Most people just
don't care about the technology used and React is rather lightweight and
will be included in Plone core soon. Therefore we choosed React (even
though, I'm using much more Angular myself).

I think React is a good default choice today and we should rather work
on a common solution than having 10 different options (which we already
have with all the Solr and ES solution).

You can always ignore/override the search view, and we should indeed
document that. Though, we should provide a sane default way of doing
this (as we used to do before).

I know this is not consensus in the Plone community, but plain Browser
views are not an option any longer (for me personally). It is 2016 and
we should start to realize that...

Am 16.08.2016 um 10:57 schrieb Tom Gross:

**Comment by gforcada** [comment]
For me a compromise would be best, i.e: have a sane and robust API with a nice 2016 UI that can be easily (i.e. documented) replaced for something else if that's your wish.

Keeping both API and UI in the same repository does indeed adds *noise* if you are only interested in one part, but knowing that you are breaking the other part is probably more important than knowing it 4 months afterwards.

**Comment by tisto** [comment]
I gave up on the separate project option basically because it would be very very hard to test. We would have tests for a plain browser view that nobody uses and we would not detect if a change in c.solr breaks the frontend.

**Comment by tomgross** [comment]
@tisto I disagree. We return a content-listing, which is well defined in Plone. As long as the backend / API / ... thing (Solr / ElasticSearch / ZCatalog / whatever) respects this Interface it does its job perfectly well. And complying with IContentListing can be easily tested from both sides.

**Comment by tisto** [comment]
@tomgross we used to have a lot of doctests in collective.solr that we had to replace with Robot tests (we still don't have tests for everything). We could write tests on all kind of levels (API, UI, Acceptance) but bottom line is nobody will do that. We are still lacking lots of tests for the current master and as long as nobody steps up to work on those I don't see how we could handle even more complexity here. 

collective.solr was always an out-of-the-box solution and providing a search UI as well. Though, we did not provide a basic browser view that worked for Plone 4 (or Plone 5). I went ahead and implemented a working solution, because I had a client that payed for that. As long as nobody steps up and implements another option that works better any discussion on that matter is useless in my opinion.

We are just talking about a default option that can be overridden at any point, we are not forcing anything on people and we are just using what has been accepted in Plone 5.1 already. If we have proper tests and an alternative implementation in place we can discuss if it is worth the effort to split packages...

I'm happy to discuss this matter in a hangout or phone call if you wish. That sometimes works better than github discussions. :)
